x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/StreamingHttpResultPublisher.java

```diff
@@ -0,0 +1,232 @@
+    // used to control the state of this publisher (Apache) and its interaction with its subscriber
+    private final AtomicBoolean isDone = new AtomicBoolean(false);
+    private final AtomicBoolean subscriptionCanceled = new AtomicBoolean(false);
+    private volatile Flow.Subscriber<? super HttpResult> subscriber;
+
+    private final RequestBasedTaskRunner taskRunner;
+    private final AtomicBoolean pendingRequest = new AtomicBoolean(false);
+    private final Deque<Runnable> queue = new ConcurrentLinkedDeque<>();
+    private final AtomicReference<CompletableFuture<Void>> threadRunner = new AtomicReference<>(CompletableFuture.completedFuture(null));
+
+    // used to control the flow of data from the Apache client, if we're producing more bytes than we can consume then we'll pause
+    private final AtomicLong bytesInQueue = new AtomicLong(0);
+    private final Object ioLock = new Object();
+    private volatile IOControl savedIoControl;
+
+    StreamingHttpResultPublisher(ThreadPool threadPool, HttpSettings settings, ActionListener<Flow.Publisher<HttpResult>> listener) {
+        this.settings = Objects.requireNonNull(settings);
+        this.listener = Objects.requireNonNull(listener);
+
+        this.taskRunner = new RequestBasedTaskRunner(new OffloadThread(), threadPool, UTILITY_THREAD_POOL_NAME);
+    }
+
+    @Override
+    public void responseReceived(HttpResponse httpResponse) throws IOException {
+        this.response = httpResponse;
+        var firstResponse = HttpResult.create(settings.getMaxResponseSize(), response);
+        this.queue.offer(() -> subscriber.onNext(firstResponse));
+        this.listener.onResponse(this);
+    }
+
+    @Override
+    public void subscribe(Flow.Subscriber<? super HttpResult> subscriber) {
+        if (this.subscriber != null) {
+            subscriber.onError(new IllegalStateException("Only one subscriber is allowed for this Publisher."));
+            return;
+        }
+
+        this.subscriber = subscriber;
+        subscriber.onSubscribe(new HttpSubscription());
+    }
+
+    @Override
+    public void consumeContent(ContentDecoder contentDecoder, IOControl ioControl) throws IOException {
+        // if the subscriber canceled us, tell Apache
+        if (subscriptionCanceled.get()) {
+            ioControl.shutdown();
+            return;
+        }
+
+        var buffer = new SimpleInputBuffer(4096);
+        var consumed = buffer.consumeContent(contentDecoder);
+        var allBytes = new byte[consumed];
+        buffer.read(allBytes);
+
+        // we can have empty bytes, don't bother sending them
+        if (allBytes.length > 0) {
+            queue.offer(() -> {
+                subscriber.onNext(new HttpResult(response, allBytes));
+
+                if (savedIoControl != null) {
+                    var maxBytes = settings.getMaxResponseSize().getBytes() * 0.5;
+                    if (bytesInQueue.updateAndGet(current -> Long.min(0, current - allBytes.length)) <= maxBytes) {
```

Comment:
Shouldn't the `updateAndGet(current -> Long.min(0, current - allBytes.length)` part always occur whether `savedIoControl !=  null` or not. ie. outside of the conditional

Reply:
Yes, this is a bug - I added a unit test to verify it and fixed it